cpp/daal/src/externals/service_service_ref.h

```diff
@@ -0,0 +1,179 @@
+        memcpy(dest, src, srcSize);
+        return 0;
+        // TODO: safe funtion
+        // return memcpy_s(dest, destSize, src, srcSize);
+    }
+
+    static int serv_memmove_s(void * dest, size_t destSize, const void * src, size_t smax)
+    {
+        if (destSize < smax) return static_cast<int>(ENOMEM);
+        memmove(dest, src, smax);
+        return 0;
+        // TODO: safe funtion
+        // return memmove_s(dest, destSize, src, smax);
+    }
+
+    static int serv_get_ht()
+    {
+        // TODO: real detection of hypertheading
+        return 1;
+    }
+
+    static int serv_get_ncpus()
+    {
+        // TODO: detection of npus
+        return 1;
+    }
+
+    static int serv_get_ncorespercpu()
+    {
+        // TODO: detection of ncores per cpu
+        return 1;
+    }
+
+    // TODO: The real call should be delegated to a backend library if the option is supported
+    static int serv_set_memory_limit(int type, size_t limit)
+    {
+        return 0;
+        // Old one - just to see what the method is for
+        // return fpk_serv_set_memory_limit(type, limit);
+    }
+
+    static size_t serv_strnlen_s(const char * src, size_t slen)
+    {
+        size_t i = 0;
+        for (; (i < slen) && (src[i] != '\0'); ++i)
+            ;
+        return i;
+    }
+
+    static int serv_strncpy_s(char * dest, size_t dmax, const char * src, size_t slen)
+    {
+        if (dmax < slen + 1) return static_cast<int>(ENOMEM);
+        strncpy(dest, src, slen);
+        return 0;
+        // TODO: safe funtion
+        // return strncpy_s(dest, dmax, src, slen);
+    }
+
+    static int serv_strncat_s(char * dest, size_t dmax, const char * src, size_t slen)
+    {
+        if (dmax < slen + 1) return static_cast<int>(ENOMEM);
+        strncat(dest, src, slen);
+        return 0;
+        // TODO: safe funtion
+        // return strncat_s(dest, dmax, src, slen);
+    }
+
+    // TODO: not a safe function - no control for the input buffer end
+    static double serv_string_to_double(const char * nptr, char ** endptr)
+    {
+        const char * cur = nptr;
+        for (; isdigit(*cur) || *cur == '-' || *cur == 'e' || *cur == 'E' || *cur == '.'; ++cur)
+            ;
+        if (endptr) *endptr = const_cast<char *>(cur);
+        size_t size = cur - nptr;
+        // TODO replace with static buffer
+        char * buffer = static_cast<char *>(malloc(size + 1));
+        for (size_t i = 0; i < size; ++i) buffer[i] = nptr[i];
+        buffer[size] = '\0';
+        double val   = atof(buffer);
+        free(buffer);
+        return val;
+    }
+
+    static float serv_string_to_float(const char * nptr, char ** endptr) { return static_cast<float>(serv_string_to_double(nptr, endptr)); }
+
+    // TODO: not a safe function - no control for the input buffer end
+    static int serv_string_to_int(const char * nptr, char ** endptr)
+    {
+        const char * cur = nptr;
+        for (; isdigit(*cur) || *cur == '-'; ++cur)
```

Comment:
This can set endptr to past where the number was consumed, and doesn't cover the case of scientific notation, e.g. 3e-5- would have 3 as the floating point number, where it should be 3 x 10^5, and if you also allow the use of 'e', you would read the final '-', which is no longer part of the string. strtol is the function you want to use here.